### PR TITLE
DBCON-298: Add SSL/TLS support for Oracle Connection Provider

### PR DESCRIPTION
**Pull request by juldiazmule**
This PR works in conjunction with DBCLI-19 (https://github.com/mulesoft/mule-db-client/pull/105)

### CODE REVIEW

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java [hunk 1]
```diff
@@ -76,9 +82,27 @@
...
+  @DisplayName("TLS Context")
+  @Summary("The TLS factory used to create TLS secured connections")
+  private TlsContextFactory tlsContextFactory;
+
   @Override
   public String getUrl() {
-    return generateUrl();
+    StringBuilder buf = new StringBuilder(JDBC_URL_PREFIX);
+
+    if (tlsContextFactory != null) {
```
Comment (konangc87) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:101]: ternary operator =)

Comment (konangc87) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:101]: ternary operator =)

**Comment by emariotti3** [comment]
Shouldn't tlsContextFactory be initialized at some point?

File: src/main/java/org/mule/extension/db/internal/domain/connection/derby/DerbyConnectionParameters.java [hunk 2]
```diff
@@ -117,4 +118,9 @@ public String getPassword() {
   public String getUser() {
     return null;
   }
+
+  @Override
+  public java.util.Optional<TlsContextFactory> getTlsContextFactory() {
+    return java.util.Optional.empty();
```
Comment (nicomz) [inline src/main/java/org/mule/extension/db/internal/domain/connection/derby/DerbyConnectionParameters.java:124]: Does derby support TLS? Is this change aligned that in a near future we may be adding TLS Support for every connection provider?

File: src/main/java/org/mule/extension/db/internal/domain/connection/mysql/MySqlConnectionParameters.java [hunk 3]
```diff
@@ -112,6 +113,11 @@ public String getUser() {
     return user;
   }
 
+  @Override
+  public java.util.Optional<TlsContextFactory> getTlsContextFactory() {
+    return java.util.Optional.empty();
```
Comment (nicomz) [inline src/main/java/org/mule/extension/db/internal/domain/connection/mysql/MySqlConnectionParameters.java:118]: Same as Derby

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java [hunk 4]
```diff
@@ -114,6 +136,40 @@ private String generateUrl() {
       buf.append("/");
       buf.append(serviceName);
     }
+
     return buf.toString();
   }
+
+  private String generateSecureUrl(StringBuilder buf) {
+    buf.append("(DESCRIPTION=");
```
Comment (nicomz) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:144]: The StringBuffer should be refactored in an Oracle TNS Entry builder class.

File: src/main/java/org/mule/extension/db/internal/domain/connection/sqlserver/SqlServerConnectionParameters.java [hunk 5]
```diff
@@ -138,6 +139,11 @@ public String getUser() {
     return user;
   }
 
+  @Override
+  public java.util.Optional<TlsContextFactory> getTlsContextFactory() {
```
Comment (nicomz) [inline src/main/java/org/mule/extension/db/internal/domain/connection/sqlserver/SqlServerConnectionParameters.java:143]: Same as Derby

**Review by nicomz — CHANGES_REQUESTED** [review]

File: src/main/java/org/mule/extension/db/internal/domain/connection/derby/DerbyConnectionParameters.java (hunk 2 above)
Comment (juldiazmule) [reply src/main/java/org/mule/extension/db/internal/domain/connection/derby/DerbyConnectionParameters.java:124]: No idea, but if it does by implementing this method we would offer the feature.

File: src/main/java/org/mule/extension/db/internal/domain/connection/mysql/MySqlConnectionParameters.java (hunk 3 above)
Comment (juldiazmule) [reply src/main/java/org/mule/extension/db/internal/domain/connection/mysql/MySqlConnectionParameters.java:118]: No idea, but if it does by implementing this method we would offer the feature.

File: src/main/java/org/mule/extension/db/internal/domain/connection/sqlserver/SqlServerConnectionParameters.java (hunk 5 above)
Comment (juldiazmule) [reply src/main/java/org/mule/extension/db/internal/domain/connection/sqlserver/SqlServerConnectionParameters.java:143]: No idea, but if it does by implementing this method we would offer the feature.

**Comment by juldiazmule** [comment]
> Shouldn't tlsContextFactory be initialized at some point?

yes, it should.

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java (hunk 4 above)
Comment (juldiazmule) [reply src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java]: Added builder class, very basic. Eventually the user will be allowed to add properties to a map and the builder will work with it

File: src/main/java/org/mule/extension/db/internal/domain/connection/derby/DerbyConnectionParameters.java (hunk 2 above)
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/derby/DerbyConnectionParameters.java:124]: static import

File: src/main/java/org/mule/extension/db/internal/domain/connection/mysql/MySqlConnectionParameters.java (hunk 3 above)
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/mysql/MySqlConnectionParameters.java:118]: static import

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java [hunk 6]
```diff
@@ -96,16 +117,16 @@ public String getUser() {
     return user;
   }
 
-  private String generateUrl() {
-    StringBuilder buf = new StringBuilder(JDBC_URL_PREFIX);
+  @Override
+  public java.util.Optional<TlsContextFactory> getTlsContextFactory() {
+    return java.util.Optional.ofNullable(tlsContextFactory);
```
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:122]: static import

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java [hunk 7]
```diff
@@ -114,6 +135,90 @@ private String generateUrl() {
...
+        .withServiceName(serviceName)
+        .build(buf);
+  }
+
+  private void checkInstanceAndServiceName() {
+    if (instance != null && serviceName != null) {
+      String errorMessage = "Instance (SID) : [" + instance + "] and Service Name : [" + serviceName
+          + "] were provided at the same time, please configure only one";
+
+      throw new ModuleException(errorMessage, CONNECTIVITY, new ConnectionException(errorMessage));
```
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:157]: why not just throwing the ConnectionException directly?

File: src/main/java/org/mule/extension/db/internal/domain/connection/sqlserver/SqlServerConnectionParameters.java [hunk 8]
```diff
@@ -138,6 +139,11 @@ public String getUser() {
...
+    return java.util.Optional.empty();
```
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/sqlserver/SqlServerConnectionParameters.java:144]: static import

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java (hunk 7 above)
Comment (juldiazmule) [reply src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:158]: The signature does not have an Exception **_throw_** clause, so the method relied on ModuleException (a RuntimeException, an unchecked exception).
Is there any way to get around this?

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java [hunk 9]
```diff
@@ -114,6 +136,90 @@ private String generateUrl() {
...
+    }
+  }
+
+  /**
+   *
+   * Builder class to facilitate the creation of Oracle's URLs that use TNS Entries
+   *
+   */
+  public static final class OracleTNSEntryBuilder {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:167]: what's the purpose of this being public if the constructor is private and there's no method returning an instance of it. This seems better fit as a package scoped top level class which can be unit tested

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java [hunk 10]
```diff
@@ -114,6 +136,90 @@ private String generateUrl() {
...
+      this.instanceName = instanceName;
+      return this;
+    }
+
+    public OracleTNSEntryBuilder withServiceName(String serviceName) {
+      this.serviceName = serviceName;
+      return this;
+    }
+
+    public String build(StringBuilder buf) {
```
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:202]: bad use of OOO. This receives a buffer of unknown state and content, mutates it and then returns the to string of that buf. That means that:

* you don't know who is invoking this method nor what's happening after, meaning the buffer will potentially receive `toString()` twice 
* You don't know the state of the buffer , which means you don't really know what is this class returning
* This class is operating under the unspoken assumption that this is the termination of the url building process.

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleDbConnectionProvider.java [hunk 11]
```diff
@@ -54,14 +57,18 @@
...
   private OracleConnectionParameters oracleConnectionParameters;
 
-
   Map<String, Map<Integer, ResolvedDbType>> resolvedDbTypesCache = new ConcurrentHashMap<>();
 
+  @Override
+  public void initialise() throws InitialisationException {
+    super.initialise();
+
+    LifecycleUtils.initialiseIfNeeded(oracleConnectionParameters.getTlsContextFactory());
```
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleDbConnectionProvider.java:69]: static import

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleDbConnectionProvider.java [hunk 12]
```diff
@@ -96,4 +103,31 @@ protected DbConnection createDbConnection(Connection connection) throws Exceptio
...
+      return false;
+    }
+
+    if (!(o instanceof OracleDbConnectionProvider)) {
+      return false;
+    }
+
+    OracleDbConnectionProvider that = (OracleDbConnectionProvider) o;
+
+    return new EqualsBuilder()
```
Comment (marianogonzalez) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleDbConnectionProvider.java:122]: replace with `Objects.equals(Object, Object)`

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java (hunk 7 above)
Comment (marianogonzalez) [reply src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleConnectionParameters.java:158]: I think the actual problem is that you're putting connectivity logic in a class which is only meant to hold parameters. I think the problem is that this logic actually belongs in the connection provider.

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

**Review by marianogonzalez — DISMISSED** [review]

**Review by konangc87 — COMMENTED** [review]
LGTM

**Review by konangc87 — APPROVED** [review]

**Review by nicomz — APPROVED** [review]

**Pull request merged**